src/rgw/rgw_admin.cc

```diff
@@ -2609,6 +2613,78 @@ static int trim_sync_error_log(int shard_id, const ceph::real_time& start_time,
   // unreachable
 }
 
+static int purge_bucket_bi(RGWRados *store,
+                           const RGWBucketInfo& bucket_info,
+                           const rgw_bucket& bucket)
+{
+  int max_shards = (bucket_info.num_shards > 0 ? bucket_info.num_shards : 1);
+  for (int i = 0; i < max_shards; i++) {
+    RGWRados::BucketShard bs(store);
+    int shard_id = (bucket_info.num_shards > 0  ? i : -1);
+    int ret = bs.init(bucket, shard_id);
+    if (ret < 0) {
+      cerr << "ERROR: bs.init(bucket=" << bucket << ", shard=" << shard_id << "): " << cpp_strerror(-ret) << std::endl;
+      return -ret;
+    }
+
+    ret = store->bi_remove(bs);
+    if (ret < 0) {
+      cerr << "ERROR: failed to remove bucket index object: " << cpp_strerror(-ret) << std::endl;
+      return -ret;
+    }
+  }
+  return 0;
+}
+
+static int cleanup_stale_bi(RGWRados *store,
+                            const string& tenant,
+                            const string& bucket_name,
+                            const rgw_bucket& orig_bucket,
+                            std::string&& stale_index)
+{
+  while (stale_index != orig_bucket.bucket_id){
+    RGWBucketInfo stale_bucket_info;
+    rgw_bucket stale_bucket;
+
+    int ret = init_bucket(tenant, bucket_name, stale_index, stale_bucket_info, stale_bucket);
+    if (ret < 0) {
+      cerr << "ERROR: could not init bucket: " << cpp_strerror(-ret) << std::endl;
+      return ret;
+    }
+    std::cout << "purging index: " << stale_index << std::endl;
+    ret = purge_bucket_bi(store, stale_bucket_info, stale_bucket);
+    if (ret < 0) {
+      return ret;
+    }
+    // continue purging subsequently resharded buckets
+    stale_index = std::move(stale_bucket_info.new_bucket_instance_id);
```

Comment:
that's clever to chain this based on `new_bucket_instance_id`. i don't know the details of the repeated-reshard bug, but it sounded like it was continuously resharding the same bucket instance, rather than the most recently resharded one - so i'm not sure it would produce chains like this?